### ISSUE TITLE
False positive for TCH002 when type used in a method

### ISSUE THREAD

**Issue by last-partizan** [astral-sh/ruff#13926]
I'm using DRF and drf-spectacular, and it uses runtime type introspection for some cases (like `serializers.SerializerMethodField()`. And i have no way to tell `ruff` to not move those types into TYPE_CHECKING block.

I tried `runtime-evaluated-base-classes`, but looks like it only works for class fields, not class methods.

```
> ruff --version
ruff 0.7.1
```

```toml
[tool.ruff]
target-version = "py39"

[tool.ruff.lint]
select = ["TCH"]

[tool.ruff.lint.flake8-type-checking]
runtime-evaluated-base-classes = [
	"rest_framework.serializers.Serializer",
	"rest_framework.serializers.ModelSerializer",
]
```

```py
from __future__ import annotations

from rest_framework import serializers

from app.models import Book


class BookSerializer(serializers.Serializer):
    name = serializers.SerializerMethodField()

    def get_name(self, instance: Book) -> str:
        return instance.name
```

```sh
> ruff check --select TCH project/serializers.py
project/serializers.py:5:24: TCH002 Move third-party import `app.models.Book` into a type-checking block
  |
3 | from rest_framework import serializers
4 |
5 | from app.models import Book
  |                        ^^^^ TCH002
  |
  = help: Move into type-checking block
```

Workaround for this is probably using `runtime-evaluated-base-classes` and adding `instance: MyClass` for affected serializers. But would be nice to have proper solution.

**Comment by MichaReiser** [comment]
I'm not familiar with any of the frameworks that you mentioned in the summary and not knowing `Book`s definition makes it a bit hard to follow the example. Could you explain me where the code depends on runtime type annotations and how Book'` defined (is it just a django model or is there more?)

**Comment by last-partizan** [comment]
Yes, `Book` is just a Django model. It's not really important, can be anything.

`djangorestframework` is used to define API/serializers, and `drf_spectacular` is used to build `schema.yml` from it.

And for `name` field, it uses `get_type_hints` from `typing` module, to know what it returns.

```python
# Partial traceback
    return append_meta(self._map_response_type_hint(method), meta)
# ...   .../site-packages/drf_spectacular/openapi.py:1119: in _map_response_type_hint
    hint = get_override(method, 'field') or get_type_hints(method).get('return')
```

I can make a full repo with reproduction, but it's actually boils down to this:

```python
from __future__ import annotations
from typing import TYPE_CHECKING, get_type_hints

if TYPE_CHECKING:
    # We're just using this as example, instead of 'from app.models import Book'
    from collections.abc import MutableMapping


class Serializer: ...


class SerializerMethodField: ...


class BookSerializer(Serializer):
    name = SerializerMethodField()

    def get_name(self, instance: MutableMapping) -> str: ...


# drf_spectacular calls get_type_hints on a method to get return type
get_type_hints(BookSerializer.get_name)
```

I hope that's enough, but if not, i cam make full repo with example.

**Comment by Avasam** [comment]
If you use libraries that need to inspect type annotations at runtime like `djangorestframework` or `pydantic`, then I don't think `TCH001`/`TCH002`/`TCH003` is made for you.

**Comment by MichaReiser** [comment]
@AlexWaygood would you mind taking a look at this. This is definitely beyond my Python knowledge 😅

**Comment by last-partizan** [comment]
> I think what's being asked here is for a way to specify that all subclasses of a certain type have all their method annotations not be picked up by TCH001/TCH002/TCH003. That way one could say that if the type is used as an annotation in a Serializer subclass, to not remove it.

Yes, that's exactly what's needed.

**Comment by MichaReiser** [comment]
Somewhat related #13713

**Comment by stiangrim** [comment]
We are struggling with this as well (Ruff telling me to move stuff into TYPE_CHECKING blocks, which breaks drf-spectacular schema). Any updates? 🙏

**Comment by johannesloibl** [comment]
This is sooo annoying...
